File: build-scripts/android_build.py
#!/usr/bin/env python

import json
import os
import subprocess
import sys

from config import Config
from lxml import etree
import shell

def install_android():
    os.system('brew install android')
    android_home = subprocess.getoutput('brew --prefix android')
    os.environ['ANDROID_HOME'] = android_home
    print('export ANDROID_HOME=%s' % android_home)

    names = [
             'platform-tools',
             'tools',
             'android-21',
             'android-22',
             'extra-google-m2repository',
             'extra-android-support',
             'extra-android-m2repository',
             'build-tools-21.1.2',
             'build-tools-22.0.1'
             ]
    for name in names:
        print('Installing', name)
        shell.cmd('echo y | android update sdk --no-ui --all --filter %s > /dev/null' % name)

def keystore():
    store = Config.file('android', 'keystore')
    print('Using keystore:', store)
    build = {
             'keystore': os.path.abspath(store),
             'storePassword': Config.get('platforms.android.keystore.PASSWORD'),
             'alias': Config.get('platforms.android.keystore.ALIAS'),
             'password': Config.get('platforms.android.keystore.ALIAS_PASSWORD')
             }
    target = os.path.join('platforms', 'android', 'build.json')
    with open(target, mode='w') as file:
        json.dump({'android': {'release': build}}, file, indent=4)

def build_num():
    num = os.environ['BUILD_NUM']
    print('Setting build_num', num)
    target = 'config.xml'
    with open(target, mode='rb') as file:
        elem = etree.fromstring(file.read())
    elem.attrib['android-versionCode'] = num
    with open(target, mode='wb') as file:
        file.write(etree.tostring(elem, encoding='utf-8', xml_declaration=True))

def build():
    mode = os.environ['BUILD_MODE']
    print('Building by cordova', mode)
    dir = os.path.join('platforms', 'android')
    multi = 'true'
    if mode != "release" and mode != "beta":
        multi = 'false'
    key = 'cdvBuildMultipleApks'
    target = os.path.join(dir, 'gradle.properties')
    def read_lines():
        if os.path.exists(target):
            with open(target, mode='r') as file:
                lines = file.readlines()
                lines = filter(lambda a: not key in a, lines)
                return map(lambda a: a.rstrip(), lines)
    lines = list(read_lines())
    lines.append('%s=%s' % (key, multi))
    with open(target, mode='w') as file:
        file.write('\n'.join(lines) + '\n')
    print('Add', target, ':', key, '=', multi)
    shell.cmd('cordova build android --release --buildConfig=%s' % os.path.join(dir, 'build.json'))

def deploy():
    import android_deploy
    android_deploy.all()

def all():
    print('Building Android')
    install_android()
    keystore()
    build_num()
    build()
    deploy()

if __name__ == "__main__":
    Config.load()

    action = sys.argv[1]
    if action == "install":
        install_android()
    elif action == "keystore":
        keystore()
    elif action == "build_num":
        build_num()
    elif action == "build":
        build()
    elif action == "deploy":
        deploy()

File: fastlane/actions/android_build.rb
module Fastlane
  module Actions
    class AndroidBuildAction < Action
      def self.run(params)
        update_sdk(params[:sdks] || [])
        build_num

        config_file = Dir.chdir(File.join('platforms', 'android')) do
          multi_apks(params[:multi_apks])
          keystore(params[:keystore])
        end

        sh("cordova build android --release --buildConfig=#{config_file}")
      end

      def self.update_sdk(names)
        begin
          content = sh('android list sdk --extended')
        rescue
          system('brew install android')
          ENV['ANDROID_HOME'] = sh('brew --prefix android')
          retry
        end

        availables = content.split("\n").map { |line|
          /^id: +\d+ or "(.*)"$/.match line
        }.compact.map { |m| m[1] }

        names.each do |name|
          puts "Checking SDK: #{name}"
          availables.select { |x| x.start_with?(name) }.each do |key|
            puts "Installing SDK #{key}"
            system("echo y | android update sdk --no-ui --all --filter #{key} | grep Installed")
          end
        end
      end

      def self.build_num
        v = ENV["BUILD_NUM"]
        if v != nil then
          num = "#{v}00"
          target = 'config.xml'
          puts "Setting build number '#{num}' to #{target}"

          require 'rexml/document'
          doc = REXML::Document.new(open(target))

          doc.elements['widget'].attributes['android-versionCode'] = num
          File.open(target, 'w') do |file|
            file.puts doc
          end
        end
      end

      def self.keystore(file)
        data = {:android => {:release =>{
          :keystore => file,
          :storePassword => ENV['ANDROID_KEYSTORE_PASSWORD'],
          :alias => ENV['ANDROID_KEYSTORE_ALIAS'],
          :password => ENV['ANDROID_KEYSTORE_ALIAS_PASSWORD']
          }}}

        target = 'build.json'
        puts "Writing #{target}"
        File.open(target, 'w') do |file|
          JSON.dump(data, file)
        end

        File.absolute_path target
      end

      def self.multi_apks(multi)
        key = 'cdvBuildMultipleApks'

        target = 'gradle.properties'
        lines = File.exist?(target) ? File.readlines(target) : []

        File.open(target, 'w+') do |file|
          lines.reject { |line| line.include?(key) }.each do |line|
            file.puts line
          end
          file.puts "#{key}=#{multi}"
        end

        File.absolute_path target
      end

      #####################################################
      # @!group Documentation
      #####################################################

      def self.description
        "Android Build"
      end

      def self.available_options
        [
          FastlaneCore::ConfigItem.new(key: :sdks,
          description: "Array of sdk names",
          optional: true,
          is_string: false
          ),
          FastlaneCore::ConfigItem.new(key: :keystore,
          description: "Absolute path to keystore",
          optional: false,
          is_string: false
          ),
          FastlaneCore::ConfigItem.new(key: :multi_apks,
          description: "Boolean for build multiple apks",
          optional: false,
          is_string: false
          )
        ]
      end

      def self.authors
        ["Sawatani"]
      end

      def self.is_supported?(platform)
        platform == :android
      end
    end
  end
end
